### PR TITLE
A few tox.ini cleanups and improvements

### PR DESCRIPTION
**Pull request by brunobord**
draft, may @ewjoachim continue on this to fix the pep8 errors, etc.

### CODE REVIEW

**Comment by brunobord** [comment]
your turn @ewjoachim

**Comment by ewjoachim** [comment]
Ok, @brunobord , it's now cleaned thanks to you. I solved the makefile / tox problem by teaching runtests.py to do coverage. I also cleaned the travis config and made it so that is displays the right python version (even if it's tox who decides in the end) Should be all good now. Your opinion on the matter ?

File: .travis.yml [hunk 1]
```diff
@@ -1,15 +1,38 @@
...
-  - TOX_ENV=py27-django18
-  - TOX_ENV=py35-django18
-  - TOX_ENV=py27-django19
-  - TOX_ENV=py35-django19
-  - TOX_ENV=py27-django110
-  - TOX_ENV=py35-django110
+
+matrix:
+    include:
+      # Python version is just ofr the look on travis.
```
Comment (brunobord) [inline .travis.yml]: "for"

File: .travis.yml [hunk 2]
```diff
@@ -1,15 +1,38 @@
...
+          env: TOX_ENV=py35-django19
+
+        - python: 3.5
+          env: TOX_ENV=py35-django110
+
+        - env: TOX_ENV=linters
+
+
+before_install:
+  - pip install codecov
```
Comment (brunobord) [inline .travis.yml]: why "before install" ? why not at install time, with tox?

File: runtests.py [hunk 3]
```diff
@@ -1,24 +1,45 @@
...
-from django.test.utils import get_runner
-from django.conf import settings
-os.environ["DJANGO_SETTINGS_MODULE"] = "tests.readonly_project.settings"
-setup()
+@contextlib.contextmanager
+def cover():
+    do_coverage = "COVERAGE" in os.environ
+
+    if do_coverage:
+        import coverage
```
Comment (brunobord) [inline runtests.py:12]: an import that isn't at the top-level? flake8 doesn't say something about it?

Comment (ewjoachim) [reply runtests.py:12]: Flake8 only lints the package source and the tests, not this file. We can change this, though.

File: .travis.yml (hunk 1 above)
Comment (ewjoachim) [reply .travis.yml]: oh. yes.

File: .travis.yml (hunk 2 above)
Comment (ewjoachim) [reply .travis.yml]: no real reason, it was done this way on the sample I used, but after examining the matter, it seems it's indeed better done in install.

File: runtests.py (hunk 3 above)
Comment (ewjoachim) [reply runtests.py:12]: And there's no reason for it to crash if coverage isn't installed but we're launching tests without coverage.

Comment (ewjoachim) [reply runtests.py:12]: It's not a problem to have an import that's not at the top if it's in a function. I've added this file to flake8 but nothing is reported.

File: .travis.yml (hunk 1 above)
Comment (k4nar) [reply .travis.yml]: Wow that's a nice trick!

File: .travis.yml [hunk 4]
```diff
@@ -1,15 +1,38 @@
...
+
 install:
   - pip install tox
+
 script:
   - tox -e $TOX_ENV
+
+after_success:
+  - codecov
```
Comment (k4nar) [inline .travis.yml:34]: et donc le coverage est correctement mergé entre les différents runs ?

File: Makefile [hunk 5]
```diff
@@ -17,7 +17,7 @@ clean-pyc: ## remove Python file artifacts
 	find . -name '*~' -exec rm -f {} +
 
 lint: ## check style with flake8
-	flake8 django_readonly_field tests
+	flake8 --exclude="migrations" django_readonly_field tests
```
Comment (k4nar) [inline Makefile]: it's better to do it in the `setup.cfg`. For example your editor won't complaint about the migrations file.

File: runtests.py (hunk 3 above)
Comment (k4nar) [reply runtests.py:12]: That's perfectly valid for flake8. Note that it isn't if you are not in a block.

File: setup.py [hunk 6]
```diff
@@ -42,7 +42,8 @@
         'django_readonly_field',
     ],
     include_package_data=True,
-    install_requires=[],
+    install_requires=["Django>=1.8,<1.11"],
```
Comment (k4nar) [inline setup.py:45]: why <1.11 ?

File: .travis.yml (hunk 4 above)
Comment (ewjoachim) [reply .travis.yml:34]: (let's keep it in english :p :gb: )
It's the only remaining question I had, and I'm unsure but I think it's ok. It's not merged by coverage but by codecov directly, which recieves the report for every sub-build (see [here](https://codecov.io/gh/novafloss/django-readonly-field/commit/edb91096f8c32ca17095d556d500613767c2438e/builds) )

File: tox.ini [hunk 7]
```diff
@@ -16,3 +15,12 @@ deps =
     django-19: Django>=1.9,<1.10
     django-110: Django>=1.10,<1.11
```
Comment (k4nar) [inline tox.ini:16]: Maybe that would be a good idea to also test the last version of django in order to detect incompatibilities early.

File: .travis.yml (hunk 4 above)
Comment (ewjoachim) [reply .travis.yml:34]: (there's probably a last bit of work to do so that codecov reports correctly the name of the build)

File: setup.py (hunk 6 above)
Comment (brunobord) [reply setup.py:45]: because, django-readonly-fields runs on django==1.10, apparently

File: .travis.yml (hunk 4 above)
Comment (k4nar) [reply .travis.yml:34]: ok nice

File: setup.py (hunk 6 above)
Comment (ewjoachim) [reply setup.py:45]: Compatibility is not yet assured for django 1.11 and/or master. Could add this, though, but in a separate PR.

Comment (k4nar) [reply setup.py:45]: It's always a debate in the Python community whether or not you should "future proof" your requirements.
My stand is that you shouldn't pin your requirements unless you know there is an issue with that version. That way you don't go in the way of the people installing your lib.

File: tox.ini (hunk 7 above)
Comment (ewjoachim) [reply tox.ini:16]: Yes, but I'd do that in a separate PR :)

**Comment by ewjoachim** [comment]
If there's not objection, I'll be rebasing and merging soon. I'll add a PR to test with Django master later, and I'll remove the upper boundary on Django version in install_requires in that other PR.
